File: src/Systems/C/factory_method.c
#include <stdio.h>

typedef struct {
    void (*connect)(void);
    void (*query)(void);
} Database;

typedef Database (*FactoryMethod)(void);

static void postgres_connect(void) { puts("PostgreSQL connect"); }
static void postgres_query(void) { puts("PostgreSQL query"); }
static void mysql_connect(void) { puts("MySQL connect"); }
static void mysql_query(void) { puts("MySQL query"); }

static Database create_postgres(void) {
    Database database = { postgres_connect, postgres_query };
    return database;
}

static Database create_mysql(void) {
    Database database = { mysql_connect, mysql_query };
    return database;
}

static void use_database(FactoryMethod create_database) {
    Database database = create_database();
    database.connect();
    database.query();
}

int main(void) {
    use_database(create_postgres);
    use_database(create_mysql);
    return 0;
}

File: src/Systems/Zig/factory_method.zig
const std = @import("std");

const DatabaseAction = *const fn () void;

const Database = struct {
    connect: DatabaseAction,
    query: DatabaseAction,
};

const FactoryMethod = *const fn () Database;

fn postgresConnect() void {
    std.debug.print("PostgreSQL connect\n", .{});
}

fn postgresQuery() void {
    std.debug.print("PostgreSQL query\n", .{});
}

fn mysqlConnect() void {
    std.debug.print("MySQL connect\n", .{});
}

fn mysqlQuery() void {
    std.debug.print("MySQL query\n", .{});
}

fn createPostgres() Database {
    return .{ .connect = postgresConnect, .query = postgresQuery };
}

fn createMySql() Database {
    return .{ .connect = mysqlConnect, .query = mysqlQuery };
}

fn useDatabase(createDatabase: FactoryMethod) void {
    const database = createDatabase();
    database.connect();
    database.query();
}

pub fn main() void {
    useDatabase(createPostgres);
    useDatabase(createMySql);
}
